rig-core/src/semantic_routing.rs

```diff
@@ -0,0 +1,89 @@
+use serde::{Deserialize, Serialize};
+
+use crate::vector_store::VectorStoreIndex;
+
+pub struct SemanticRouter<V> {
+    store: V,
+    threshold: f64,
+}
+
+impl<V> SemanticRouter<V> {
+    pub fn builder() -> SemanticRouterBuilder<V> {
+        SemanticRouterBuilder::new()
+    }
+}
+
+impl<V> SemanticRouter<V>
+where
+    V: VectorStoreIndex,
+{
+    pub async fn select_route(&self, query: &str) -> Option<String> {
+        let res = self.store.top_n(query, 1).await.ok()?;
+        let (score, _, SemanticRoute { tag }) = res.first()?;
+
+        println!("{tag}, {score}");
+
+        if *score < self.threshold {
+            return None;
+        }
+
+        Some(tag.to_owned())
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SemanticRoute {
+    tag: String,
+}
+
+pub trait Router: VectorStoreIndex {
+    fn retrieve_route() -> impl std::future::Future<Output = Option<String>> + Send;
+}
+
+pub struct SemanticRouterBuilder<V> {
```

Comment:
```suggestion
#[derive(Default)]
pub struct SemanticRouterBuilder<V> {
```

Should be easier to use the `derive` here instead of an `impl` unless you want to reserve relying on `::new` usage as it could change in the future.